### ISSUE TITLE
How to identify generic types?

### ISSUE THREAD

**Issue by jamii** [ziglang/zig#8830]
(Hopefully this is not a dupe - I couldn't figure out what to search for).

Suppose I'm writing a data inspector or similar. Usually i end up with code that looks something like this:

``` zig
fn inspect(thing: anytype) void {
  const T = @TypeOf(thing);
  if (@hasDecl(T, inspect)) {
    // use override
    thing.inspect();
  } else {
   // fall back to generic version
    const ti = @typeInfo(T):
    switch (ti) {
      ...
    }
}
```

The trouble comes when I want to add special logic for existing types that I can't add methods to, like std.ArrayList. I would need to be able to write something like `if (T.wasCreatedBy(std.ArrayList))` which as far as I can tell is not currently possible. That information does exist, because it's used to name the type, but the name isn't sufficiently unique to tell which function created it. 

Perhaps `@typeParentFunction(t: type) ?fn` or similar?

**Comment by g-w1** [comment]
There has been a discussion here about this. https://zigforum.org/t/check-if-a-type-is-a-generic-type/436/5

**Comment by andrewrk** [comment]
There is no way to do this, and no plan to implement any language features that would allow you to do this.

**Comment by YohananDiamond** [comment]
I've been thinking about that too. I was doing some experiments with an interface type and I came up with [this](https://github.com/YohananDiamond/zig-poly/blob/34788f4543ed887e40b20c2cde4c4e014125ab40/src/poly.zig#L38-L47):
```zig
pub fn isBaseOf(comptime IType: type) bool {
    comptime {
        const field_name = "_InterfaceBaseType";

        if (!meta.trait.is(.Struct)(IType)) return false;
        if (!@hasDecl(IType, field_name)) return false;
        if (@TypeOf(@field(IType, field_name)) != type) return false;
        return @field(IType, field_name) == IFaceSelf;
    }
}
```
Basically I created a `_InterfaceBaseType` constant on my struct that contains the type argument used on the function that generated this struct. It's messy but it works (I might switch to jmc's hack used on the mentioned zig forum post later).
Having a builtin way of doing this would make this so much easier for these specific situations but I'm not sure if it's something that is used often, so putting a lot of effort on implementing it may be not worth it.
